pkg/lib/oauth/handler/handler_token.go

```diff
@@ -755,8 +757,15 @@ func (h *TokenHandler) handleBiometricAuthenticate(
 		return nil, err
 	}
 
-	// TODO(DEV-1404): allow specifying scopes
 	scopes := []string{"openid", oauth.FullAccessScope}
+	requestedScopes := r.Scope()
+	if len(requestedScopes) > 0 {
+		err := h.ValidateScopes(client, requestedScopes)
+		if err != nil {
+			return nil, err
+		}
+		scopes = requestedScopes
```

Comment:
If the client provides only the app-initiated-sso-to-web scope, then scopes will be missing `openid` and `full-access` scope.

And I realize we are missing the scope `offline_grant` here. `offline_grant` should be present in order to issue an offline grant. It seems that we have this problem in places where scopes are hard-coded.

Reply:
@louischan-oursky 
Lets do a validation instead of implicitly add the scopes?

Reply:
Validate the implicitly added scopes, and the explicitly provided scope.